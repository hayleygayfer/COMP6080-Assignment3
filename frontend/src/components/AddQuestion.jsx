import React from 'react';
import PropTypes from 'prop-types';
import '../App.css'
import API from '../api.js';
const api = new API('http://localhost:5005');

function AddQuestion (id) {
  const token = localStorage.getItem('token');
  const [gameData, setGameData] = React.useState('');
  const [questionString, setQuestionString] = React.useState('');
  const [questionAnswers, setQuestionAnswers] = React.useState('');
  const [timeLimit, setTimeLimit] = React.useState('');
  const [questionMedia, setQuestionMedia] = React.useState('');
  const gameId = id.input;

  const getGameQuestionsRequest = async (quizId) => {
    try {
      const request = await api.makeAPIRequest(`admin/quiz/${quizId}`, token, 'GET', '', '');
      if (request.status === 200) {
        console.log('Got Game Data');
        const data = await request.json();
        setGameData(data);
      } else throw request.status
    } catch (error) {
      alert(`Invalid Quiz Request: ${error}`);
      console.log(error);
    }
  }

  let questionEditor = [];
  const openQuestionEditor = () => {
    getGameQuestionsRequest(gameId);
    questionEditor.push(<>
      <div>
        Question: <input
          type="text"
          onChange={e => setQuestionString(e.target.value)}
          value={questionString}
        /><br/>
        Time Limit: <input
          type="text"
          onChange={e => setTimeLimit(e.target.value)}
          value={timeLimit}
        /><br/>
        Answers: <input
          type="text"
          onChange={e => setQuestionAnswers(e.target.value)}
          value={questionAnswers}
        /><br/>
        Image or Video: <input
          type="text"
          onChange={e => setQuestionMedia(e.target.value)}
          value={questionMedia}
        /><br/>
      </div>
      <button className='button' onClick={addGameQuestionsRequest(gameId)}> Submit Question </button>
    </>)
  }

  const newQuestion = {
    questionString: questionString,
    answers: questionAnswers,
    timeLimit: timeLimit,
    imgSource: questionMedia
  };

  const addGameQuestionsRequest = async (quizId) => {
    if (!gameData) return 1;
    gameData.questions.push(newQuestion);

    try {
      const request = await api.makeAPIRequest(`admin/quiz/${quizId}`, token, 'PUT', '', gameData);
      if (request.status === 200) {
        console.log('Updated Game Data');
        questionEditor = [];
      } else throw request.status
    } catch (error) {
      alert(`Invalid Update Question Request: ${error}`);
      console.log(error);
    }
  }

  return (<>
    <button className='button' onClick={openQuestionEditor()}> Open Editor </button>
    {questionEditor}
  </>);
}

AddQuestion.propTypes = {
  gameId: PropTypes.number
};

export default AddQuestion;
