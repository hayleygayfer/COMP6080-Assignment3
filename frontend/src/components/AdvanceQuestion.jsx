import React from 'react';
import '../App.css';
import API from '../api.js';
const api = new API('http://localhost:5005');

function AdvanceQuestion () {
  const token = localStorage.getItem('token');
  const gameId = localStorage.getItem('gameIdOfStartedGame');
  const sessionId = localStorage.getItem('sessionIdOfStartedGame');
  const [curQuestion, setCurQuestion] = React.useState('');
  const [questionList, setQuestionList] = React.useState('');

  const getSessionStatus = async () => {
    try {
      const request = await api.makeAPIRequest(`admin/session/${sessionId}/status`, token, 'GET', '', '');
      if (request) {
        console.log('got Session Data');
        console.log(request);
        setCurQuestion(request.position);
        setQuestionList(request.questions);
      }
    } catch (error) {
      alert(`Invalid Question Request: ${error}`);
      console.log(error);
    }
  }

  const advanceQuestionRequest = async () => {
    try {
      const request = await api.makeAPIRequest(`admin/quiz/${gameId}/advance`, token, 'POST', '', '')
      console.log(request)
      if (request.status === 200) {
        console.log('Advanced to next question');
      } else throw request.status
    } catch (error) {
      console.log(error)
    }
  }

  const progressQuiz = () => {
    console.log(`Starting Progress Quiz for gameID: ${gameId}`);
    // START QUIZ
    advanceQuestionRequest();
    getSessionStatus();
  }

  React.useEffect(() => {
    console.log('timing');
    const timer = setTimeout(() => {
      advanceQuestionRequest();
      getSessionStatus();
      console.log('advancing');
    }, questionList[curQuestion].timeLimit * 1000);
    return () => clearTimeout(timer);
  }, []);

  return (<>
    <div>
      <button className='button' onClick={() => progressQuiz()}> Start Questions </button><br/>
      Current Question: {curQuestion}
    </div>
  </>)
}

export default AdvanceQuestion;
