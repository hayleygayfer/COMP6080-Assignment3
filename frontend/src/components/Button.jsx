import React from 'react';
import PropTypes from 'prop-types';
function Button (props) {
  return <button className='button' onClick={props.onClickFunction}>{props.text}</button>
}
export default Button;
Button.propTypes = {
  text: PropTypes.string,
  onClickFunction: PropTypes.func
};
