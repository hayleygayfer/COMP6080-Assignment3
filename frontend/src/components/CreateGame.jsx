import React from 'react';
import '../App.css';
import API from '../api.js';
const api = new API('http://localhost:5000');

function CreateGame () {
  const [nameInput, setNameInput] = React.useState('');
  const [newQuizId, setNewQuizId] = React.useState('');
  const [gameJSON, setGameJSON] = React.useState('');
  const token = localStorage.getItem('token');

  // File handling
  const [selectedFile, setSelectedFile] = React.useState();
  const [isFilePicked, setIsFilePicked] = React.useState(false);

  // File submission
  const selectFile = (event) => {
    setSelectedFile(event.target.files[0]);
    setIsFilePicked(true);
  }

  const uploadFile = () => {
    if (isFilePicked === false) return 1;

    const reader = new FileReader();
    reader.onload = onReaderLoad;
    reader.readAsText(selectedFile);
  }

  function onReaderLoad (event) {
    console.log(event.target.result);
    setGameJSON(event.target.result);
  }

  /* CREATE NEW GAME SERVER REQUEST */
  const createGameRequest = async () => {
    try {
      // if name = '', throw error
      if (nameInput === '') {
        throw Error('Name can\'t be empty');
      }
      const request = await api.makeAPIRequest('admin/quiz/new', token, 'POST', '', {
        name: nameInput,
      });
      if (request.status === 200) {
        console.log('Game Created');
        // send them to dashboard
        request.json().then((data) => {
          setNewQuizId(data.quizId)
          console.log(newQuizId);
          if (gameJSON !== '') {
            if (validateGameJSON(gameJSON) === false) {
              alert('File does not match valid JSON game data structure');
            }
            loadGameDataRequest(newQuizId);
          }
        });
        alert(`Game Created! id: ${newQuizId}`);
        setNameInput('');
      } else throw request.status
    } catch (error) {
      setNameInput('');
      alert(`Invalid New Quiz Request: ${error}`);
      console.log(error);
    }
  }

  const validateGameJSON = (gameData) => {
    let game = {};
    try {
      game = JSON.parse(gameData);
    } catch (error) {
      alert(error);
      console.log(error);
      return false;
    }

    if (!game.questions) return false;
    for (let i = 0; i < game.questions.length; i++) {
      if (!game.questions[i].questionType) return false;
      if (!game.questions[i].questionString) return false;
      if (!game.questions[i].timeLimit) return false;
      if (!game.questions[i].pointValue) return false;
      if (!game.questions[i].mediaSource) return false;
      if (!game.questions[i].answers) return false;
      if (!game.questions[i].correctAnswers) return false;
    }
    if (!game.name) return false;
    if (!game.thumbnail) return false;

    return true;
  }

  const loadGameDataRequest = async (quizId) => {
    try {
      const request = await api.makeAPIRequest(`admin/quiz/${quizId}`, token, 'PUT', '', gameJSON);
      console.log(request);
      if (request) {
        console.log('Loaded Game data');
      }
    } catch (error) {
      alert(`Invalid Load Data Request: ${error}`);
      console.log(error);
    }
  }

  return (<>
    <div>
      Game name:
      <input
        type="text"
        onChange={e => setNameInput(e.target.value)}
        value={nameInput}
      /><br/>
      Upload Game Data JSON File (optional):<br/>
      <input type="file" name="file" onChange={selectFile} />
      { isFilePicked
        ? (<div>
              <p>Filename: {selectedFile.name}</p>
              <p>FileType: {selectedFile.type}</p>
            </div>
          )
        : (
            <p>Select a file to show details</p>
          )
      }
    <div>
      <button onClick={uploadFile}>Submit</button>
    </div>
    <button className='button smallButton' onClick={createGameRequest}> Create New Quiz </button>
    </div>
  </>)
}

export default CreateGame;
