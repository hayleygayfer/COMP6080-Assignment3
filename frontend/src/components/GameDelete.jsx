import React from 'react';
import '../App.css'
import API from '../api.js';
const api = new API('http://localhost:5005');

function GameDelete (gameId) {
  const deleteRequest = async () => {
    try {
      const token = localStorage.getItem('token');
      await api.makeAPIRequest('admin/quiz/' + gameId.input, token, 'DELETE', '', '');
    } catch (error) {
    }
  }
  deleteRequest();
  return <>
    <h2> Game has been deleted </h2>
    { // LINK TO results (parameterised on the session ID)
      // localStorage.setItem('token', data.token);
    }
  </>;
}

export default GameDelete;
