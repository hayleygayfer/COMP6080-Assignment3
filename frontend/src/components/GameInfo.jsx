import React from 'react';
import PropTypes from 'prop-types';
function GameInfo (props) {
  return <>
  Title: {props.title}<br/>
  Thumbnail: {props.thumbnail ? <img src={props.thumbnail}> </img> : 'none provided' } <br/>
  </>
}
export default GameInfo;
GameInfo.propTypes = {
  title: PropTypes.string, // gameData[i].name
  thumbnail: PropTypes.string // gameData[i].thumbnail
};
