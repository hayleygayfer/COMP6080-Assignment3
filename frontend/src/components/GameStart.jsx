import React from 'react';
import '../App.css'

function GameStart (gameId) {
  console.log(gameId.input)
  const url = window.location.origin + '/play_join?game_id=' + gameId.input
  return <>
    <p>Game Id: {gameId.input}</p>
    { // LINK TO 'COPY GAME URL' .. which will be /play_join:{gameId.input}
    }
    <button className='button smallButton' onClick={ () => { navigator.clipboard.writeText(url) } }> Copy Link </button>
  </>;
}

export default GameStart;
