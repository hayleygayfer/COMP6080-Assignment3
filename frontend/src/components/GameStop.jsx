import React from 'react';
import '../App.css'
import { useHistory } from 'react-router-dom';
import API from '../api.js';
const api = new API('http://localhost:5005');

function GameStop (gameId) {
  const history = useHistory();
  const stopRequest = async () => {
    try {
      // insert token
      console.log('stopping game....')
      const token = localStorage.getItem('token');
      const request = await api.makeAPIRequest('admin/quiz/' + gameId.input + '/end', token, 'POST', '', '')
      console.log(request)
    } catch (error) {
    }
  }
  stopRequest()
  const showResults = () => {
    console.log(gameId.input)
    const path = '/game_results?game_id=' + gameId.input
    // GETS THE RIGHT URL... BUT DOESN'T REFRESH
    history.push(path)
    window.location.reload();
  }
  return <>
    <h2> Would you like to see the results? </h2>
    { // LINK TO results (parameterised on the session ID)
      // localStorage.setItem('token', data.token);
    }
    <button className='button smallButton' onClick={showResults}> Yes </button>
  </>;
}

export default GameStop;
