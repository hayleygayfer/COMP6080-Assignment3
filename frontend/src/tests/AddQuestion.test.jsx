import { shallow } from 'enzyme';
import * as React from 'react';
import { AddQuestion } from '../components/AddQuestion';
describe('AddQuestion', () => {
  it('triggers onClick event handler when clicked', () => {
    const mockCallBack = jest.fn();
    const buttonData = shallow((<AddQuestion />)).first('button');
    console.log(buttonData.html())
    const button = shallow((<buttonData onClick={mockCallBack}>Ok!</buttonData>));
    button.simulate('click');
    console.log(buttonData.html())
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  // has questionEditor appears when add question button clicked
  it('questionEditor appears when "AddQuestion" button clicked', () => {
    const wrapper = shallow((<AddQuestion />));
    wrapper.find('button').simulate('click');
    console.log(wrapper.html())
    expect(wrapper.html()).toEqual('<button class="button"> Open Editor </button><div>Question: <input type="text" value=""/><br/>Time Limit: <input type="text" value=""/><br/>Answers: <input type="text" value=""/><br/>Image or Video: <input type="text" value=""/><br/></div><button class="button"> Submit Question </button>');
  });

  // doesn't have questionEditor when not clicked
  it('initially only returns a button', () => {
    const wrapper = shallow(<AddQuestion />);
    expect(wrapper.html()).toEqual('<button class="button"> Open Editor </button>');
  });
});
