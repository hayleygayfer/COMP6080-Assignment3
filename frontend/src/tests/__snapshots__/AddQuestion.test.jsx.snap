// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddQuestion renders correctly when click happens (snapshot) 1`] = `
<button
  className="button"
  onClick={[Function]}
>
   Open Editor 
</button>
`;

exports[`AddQuestion renders correctly when no click (snapshot) 1`] = `
<button
  className="button"
  onClick={[Function]}
>
   Open Editor 
</button>
`;
